clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/InfinispanBeanManagerFactoryServiceConfigurator.java

```diff
@@ -23,85 +23,76 @@
+    private volatile Supplier<MarshallingConfigurationRepository> repository;
+    private volatile Supplier<ScheduledExecutorService> scheduler;
+    private volatile Supplier<NodeFactory<Address>> group;
+    private volatile Supplier<Registry<String, ?>> registry;
+    private volatile Supplier<CommandDispatcherFactory> dispatcherFactory;
+
+    public InfinispanBeanManagerFactoryServiceConfigurator(CapabilityServiceSupport support, String name, BeanContext context, BeanManagerFactoryServiceConfiguratorConfiguration configuration) {
+        super(context.getDeploymentUnitServiceName().append(context.getBeanName()).append("bean-manager"));
         this.support = support;
         this.name = name;
         this.context = context;
         this.configuration = configuration;
     }
 
     @Override
-    public ServiceName getServiceName() {
-        return this.context.getDeploymentUnitServiceName().append(this.context.getBeanName()).append("bean-manager");
-    }
-
-    @Override
-    public ServiceBuilder<BeanManagerFactory<I, T, TransactionBatch>> build(ServiceTarget target) {
+    public ServiceBuilder<?> build(ServiceTarget target) {
         String containerName = this.configuration.getContainerName();
         ServiceName deploymentUnitServiceName = this.context.getDeploymentUnitServiceName();
-        return target.addService(this.getServiceName(), new ValueService<>(this))
-                .addDependency(InfinispanCacheRequirement.CACHE.getServiceName(this.support, containerName, InfinispanBeanManagerFactoryBuilderFactory.getCacheName(deploymentUnitServiceName)), Cache.class, this.cache)
-                .addDependency(InfinispanRequirement.KEY_AFFINITY_FACTORY.getServiceName(this.support, containerName), KeyAffinityServiceFactory.class, this.affinityFactory)
-                .addDependency(deploymentUnitServiceName.append("marshalling"), MarshallingConfigurationRepository.class, this.repository)
-                .addDependency(deploymentUnitServiceName.append(this.name, "expiration"), ScheduledExecutorService.class, this.scheduler)
-                .addDependency(ClusteringRequirement.COMMAND_DISPATCHER_FACTORY.getServiceName(this.support, containerName), CommandDispatcherFactory.class, this.dispatcherFactory)
-                .addDependency(ClusteringCacheRequirement.REGISTRY.getServiceName(this.support, containerName, BeanManagerFactoryBuilderConfiguration.CLIENT_MAPPINGS_CACHE_NAME), Registry.class, this.registry)
-                .addDependency(ClusteringCacheRequirement.GROUP.getServiceName(this.support, containerName, BeanManagerFactoryBuilderConfiguration.CLIENT_MAPPINGS_CACHE_NAME), NodeFactory.class, this.group)
-                .setInitialMode(ServiceController.Mode.ON_DEMAND)
-        ;
-    }
-
-    @Override
-    public BeanManagerFactory<I, T, TransactionBatch> getValue() {
-        return new InfinispanBeanManagerFactory<>(this);
+        ServiceBuilder<?> builder = target.addService(this.getServiceName());
```

Comment:
FYI (and I don't think you need to change this patch because of this) but the intent of separating the service instance out of the `addService` call was so that the dependency `Supplier` and dependent `Consumer` instances could all be set on final fields when the service is constructed.  Then the `start` method would construct the value(s) based on the dependency values, and "outject" them into the consumers.